include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,406 @@
+                    const arma::vec& lowerBound,
+                    const arma::vec& upperBound) :
+    populationSize(populationSize),
+    maxGenerations(maxGenerations),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    epsilon(epsilon),
+    lowerBound(lowerBound),
+    upperBound(upperBound)
+{ /* Nothing to do here. */ }
+
+//! Optimize the function.
+template<typename MatType,
+         typename... ArbitraryFunctionType,
+         typename... CallbackTypes>
+typename MatType::elem_type NSGA2::Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
+                                     MatType& iterate,
+                                     CallbackTypes&&... callbacks)
+{
+  // Make sure for evolution to work at least four candidates are present.
+  if (populationSize < 4 && populationSize % 4 != 0)
+  {
+    throw std::logic_error("NSGA2::Optimize(): population size should be at"
+        " least 4, and, a multiple of 4!");
+  }
+
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+
+  numObjectives = sizeof...(ArbitraryFunctionType);
+  numVariables = iterate.n_rows;
+
+  // Cache calculated objectives.
+  std::vector<arma::Col<ElemType> > calculatedObjectives;
+  // Pre-allocate space for the calculated objectives.
+  calculatedObjectives.resize(populationSize);
+
+  // Population size reserved to 2 * populationSize + 1 to accommodate
+  // for the size of intermediate candidate population.
+  std::vector<MatType> population;
+  population.reserve(2 * populationSize + 1);
+
+  // Pareto fronts, initialized during non-dominated sorting.
+  std::vector<std::vector<size_t> > fronts;
+  // Initialised in CrowdingDistanceAssignment.
+  std::vector<double> crowdingDistance;
+  // Initialised during non-dominated sorting.
+  std::vector<size_t> ranks;
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // Generate the population based on a uniform distribution around the given
+  // starting point.
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    population.push_back(arma::randu<MatType>(iterate.n_rows,
+        iterate.n_cols) - 0.5 + iterate);
+  }
+
+  Info << "NSGA2 initialized successfully. Optimization started." << std::endl;
+
+  // Evaluate the fitness before optimization.
+  for (size_t i = 0; i < population.size(); i++)
+    calculatedObjectives[i] = arma::Col<ElemType>(numObjectives, arma::fill::zeros);
+  EvaluateObjectives(population, objectives, calculatedObjectives);
+
+  // Iterate until maximum number of generations is obtained.
+  terminate |= Callback::BeginOptimization(*this, objectives, iterate, callbacks...);
+
+  for (size_t generation = 1; generation <= maxGenerations && !terminate; generation++)
+  {
+    Info << "NSGA2: iteration " << generation << "." << std::endl;
+    terminate |= Callback::StepTaken(*this, objectives, iterate, callbacks...);
```

Comment:
Do we have to handle `terminate` here?  (And for the other callbacks too?)

Reply:
Good point, usually we take a step at the end, so the for loop will terminate right after that, but in this case we don't.

Reply:
Ahhhh, good point, I overlooked the `!terminate` in the for loop.  That's what all the other optimizers do, so I don't see a problem with just doing that here.  Maybe we could open some other issue to handle that if we wanted.